### ISSUE TITLE
restore symmetry through quantum number projection

### ISSUE THREAD

**Issue by dambuck** [netket/netket#510]
Hello!

I was wondering if something like this https://arxiv.org/pdf/2009.14777.pdf is easily realizable?
One would need to take linear combinations of RBMs which are translated according to a given symmetry.

**Comment by nikita-astronaut** [comment]
Hello, @dambuck ,

yes, this not very complicated and doable within Netket.

Suppose you have two different networks for amplitude and phase. Then you need to unite them and symmetrize in the following way (here I store spatial symmetries in `symms` and also perform the M_z: s_z -> -s_z projection):

`def combine_amplitude_and_phase(*modules) -> torch.nn.Module:
    class CombiningState(torch.nn.Module):
        def __init__(self, amplitude, phase, symms):
            super().__init__()
            self.amplitude = amplitude
            self.phase = phase
            self.symms = symms  # this is the N_symm x N_spins numpy array
            
        def forward(self, x):
            ampl_log, phase = logmeanexp( \
                torch.cat([self.amplitude(x[..., symm]) for symm in self.symms] + [self.amplitude(-x[..., symm]) for symm in self.symms] , dim = 1), \
                torch.cat([self.phase(x[..., symm]) for symm in self.symms] + [self.phase(-x[..., symm]) for symm in self.symms] , dim = 1), dim = 1)
            return torch.cat([ampl_log, phase], dim=1).type(torch.DoubleTensor)
    return CombiningState(*modules)`


Here also you would need

`def logmeanexp(inputs_logampls, inputs_phases, dim=None, keepdim=False):
    """Numerically stable logsumexp.

    Returns:
        Equivalent of log(sum(exp(inputs), dim=dim, keepdim=keepdim)).
    """
    # For a 1-D array x (any array along a single dimension),
    # log sum exp(x) = s + log sum exp(x - s)
    # with s = max(x) being a common choice.
    if dim is None:
        inputs_logampls = inputs_logampls.view(-1)
        inputs_phases = inputs_phases.view(-1)
        dim = 0
    # print(inputs_logampls.size(), inputs_phases.size())

    s, _ = torch.max(inputs_logampls, dim=dim, keepdim=True)
    # print(inputs_logampls, s)
    inputs_logampls -= s
    real = torch.sum(torch.exp(inputs_logampls) * torch.cos(inputs_phases), axis = 1).unsqueeze(1)
    imag = torch.sum(torch.exp(inputs_logampls) * torch.sin(inputs_phases), axis = 1).unsqueeze(1)
    # print('realimag', real.size(), imag.size())
    logampls = torch.log(real ** 2 + imag ** 2) / 2. + s - torch.log(torch.ones(real.size(0)) * inputs_logampls.size(1)).unsqueeze(1)
    phases = torch.atan2(imag, real)


    # print('after', logampls.size(), phases.size())
    return logampls, phases`


In code, you do something like

`model_phase = NN.Net_32_dense()
model_ampl = NN.Net_32_dense()
combine_amplitude_and_phase = NN.combine_amplitude_and_phase
model_combined = combine_amplitude_and_phase(model_ampl, model_phase)
ma = nk.machine.Torch(model_combined, hilbert=hi)
`

Here I also used my network for the 32-spin model.

`class Net_32_dense(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self._dense = torch.nn.Linear(32, 128, bias=True)

    def forward(self, x):
        x = self._dense(x)
        x = torch.log(torch.cosh(x))

        return torch.sum(x, axis = 1).view(x.shape[0], 1)`

**Comment by nikita-astronaut** [comment]
This `Net_32_dense` is not exactly what is given in the paper of Nomura, since it is purely real model. However, I think on the complex realization in `pytorch` @ChenAo-Phys might comment.

**Comment by dambuck** [comment]
Thank you very much for your answer and the code examples! you say this example is real but you are using combining phase and amplitude to model complex values. Why would this model behave differently from a complex one?
Did you try your model with other systems than in the paper ?

**Comment by nikita-astronaut** [comment]
Using complex weights instead of real ones brings more expressibility to the model, thus the behaviour will somehow be different.

Yes, I have tried on other graphs, usually the performance is good.

**Comment by ChenAo-Phys** [comment]
I'm not using netket now so probably my network doesn't work... but I can show it here
```
class RBM(nn.Module):
    def __init__(self):
        super().__init__()
        self.N = 32
        self.M = 128
        self.weight = torch.randn(self.N, self.M, dtype=torch.complex64) * 0.01
        self.bias = torch.randn(self.M, dtype=torch.complex64) * 0.01

    def forward(self, input):
        theta = torch.zeros(input.shape[0], self.M, dtype=torch.complex64)
        theta.real = torch.einsum('ni,ij->nj', input, self.weight.real)
        theta.imag = torch.einsum('ni,ij->nj', input, self.weight.imag)
        theta += self.bias.unsqueeze(dim=0)
        ln_psi = theta.cosh_().log_().sum(dim=1)
        return torch.view_as_real(ln_psi)
```
It's not clear whether netket works well with complex pytorch networks. If you really want to have complex networks in netket it's better to use jax.

**Comment by femtobit** [comment]
If I remember our recent discussion, this issue should now be resolved by the inclusion of `GCNN` in v3 (which can emulate a symmetry-averaged RBM with one layer and logcosh activation), right? @chrisrothUT
